### PR TITLE
Julia binding fixes

### PR DESCRIPTION
**Pull request by rcurtin**
Many of these were suggested by @NHDaly after he spent some time playing around with the Julia bindings.  (Thanks Nathan!)

### 1. Fix binding functionality on i386 by being paranoid about the types being passed back and forth.

The bindings worked fine on 64-bit systems, but on 32-bit systems, `size_t` is not 64-bits, and I don't think we have a guarantee (or at least not one I'm willing to trust, since things can change) that Julia's `int` and `uint` types will match the C/C++ types `int` and `size_t`.  So, the code is changed so that any integer or unsigned integer being passed back and forth is 64-bits (via Julia's `int64` and `uint64` types, and C/C++'s `int64_t` and `uint64_t`).  Then, once inside C/C++, this is either casted or converted to `size_t` as necessary.

Now, the bindings appear to work in an i386 docker container just fine.

*(Edit: actually we can just use `Csize_t` for this, which in Julia is always the same as the C++ `size_t` type.  See some of the comments below.)*

### 2. Fix generated Markdown documentation for Julia docstrings.

Previously, the Julia help strings for mlpack functions would display wrong:

```
help?> mlpack.knn
  ...

  For example, the following command will calculate the 5 nearest neighbors of each point in input and store the distances in distances and the neighbors in neighbors:

  julia> using CSV julia> input = CSV.read("input.csv") julia> distances, neighbors, _ = knn(k=5, reference=input)

  The output is organized such that row i and column j in the neighbors output matrix corresponds to the index of the point in the reference set which is the j'th nearest ...
```

This was because the code snippets weren't enclosed in a Markdown code block.  That's fixed now, and displays correctly:

```
help?> mlpack.knn
  ...

  For example, the following command will calculate the 5 nearest neighbors of each point in input and store the distances in distances and the neighbors in neighbors: 

  julia> using CSV
  julia> input = CSV.read("input.csv")
  julia> distances, neighbors, _ = knn(k=5, reference=input)

  The output is organized such that row i and column j in the neighbors output matrix corresponds to the index of the point in the reference set which is the j'th nearest neighbor from the point in the query set with index i. Row j and column i in the distances output matrix
  corresponds to the distance between those two points.
```

### 3. Change name of `util` module to `_Internal`

It may not be entirely clear to Julia users, who are used to modules being named with capital letters, that `util` is an internal utility module.  So, I made this more clear by renaming it to `_Internal` and adding a docstring:

```
help?> mlpack._Internal
  mlpack._Internal

  This module contains internal implementations details of mlpack. There shouldn't be any need to go digging around in here if you're just using mlpack. (But don't let this comment discourage you if you're just curious and poking around!)
```

### 4. Add a docstring to the main `mlpack` module

Previously there was no docstring for `mlpack`:

```
help?> mlpack
search: mlpack

  No documentation found.

  No docstring found for module mlpack.
```

This is now improved:

```
help?> mlpack
search: mlpack

  mlpack

  mlpack is a fast, flexible machine learning library, written in C++, that aims to provide fast, extensible implementations of cutting-edge machine learning algorithms. This module provides those implementations as Julia functions.

  Each function inside the module performs a specific machine learning task.

  For complete documentation of these functions, including example usage, see the mlpack website's documentation for the Julia bindings:

  https://www.mlpack.org/doc/stable/julia_documentation.html

  Each function also contains an equivalent docstring; the Julia REPL's help functionality can be used to access the documentation that way.
```

### 5. Remove `REQUIRE` file

As far as I understand, this isn't actually necessary anymore, since the Julia requirements are now in `Project.toml` (as of #2246 at least).

By the way, all of the `help?>` output here is from Julia's REPL, which allows you to get documentation for any function just by typing `?` then the name of the function.

### CODE REVIEW

**Comment by NHDaly** [comment]
> and I don't think we have a guarantee (or at least not one I'm willing to trust, since things can change) that Julia's `int` and `uint` types will match the C/C++ types `int` and `size_t`.

Just fyi, if this is something you need in the future, julia's `Cssize_t ` and `Csize_t` will match the system's bit size, and those also match Julia's `Int` and `UInt`:
https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#man-bits-types-1
(Scroll down to the nice table, there.)
Apparently c's `int` is actually almost always 32-bits even on most 64-bit systems, and julia's `Cint` will match that as well. BUT anyway, your change to just specify the sizes you want is probably best.

**Comment by zoq** [comment]
Looks like the markdown build failed.

**Comment by rcurtin** [comment]
@NHDaly thanks for pointing that out, apparently I must have missed the documentation.  So I made a handful of changes to make this cleaner:

 * Always use `Csize_t` to pass `size_t` back and forth, instead of `UInt64`
 * Instead of using `Int64` and `UInt64` as types in Julia, just use `Int` and `UInt`, since this will match the C++ `int` and `size_t` types.

These are done in d730879.

@zoq thanks for pointing out the Markdown build, I think I have that fixed now in 758cb5a...

File: src/mlpack/bindings/julia/julia_util.cpp [hunk 1]
```diff
@@ -485,7 +493,7 @@ size_t* CLI_GetParamURow(const char* paramName)
  */
 size_t CLI_GetParamMatWithInfoRows(const char* paramName)
 {
-  return std::get<1>(CLI::GetParam<std::tuple<data::DatasetInfo, arma::mat>>(
+  return std::get<1>( CLI::GetParam<std::tuple<data::DatasetInfo, arma::mat>>(
```
Comment (zoq) [inline src/mlpack/bindings/julia/julia_util.cpp]: What is the reason of the added space here, to improve readability?

Comment (rcurtin) [reply src/mlpack/bindings/julia/julia_util.cpp]: Nice catch, the reason is an incorrect applied vim macro. :)  Let me find all errors like this and fix them. :+1:

**Pull request merged**